crates/red_knot_python_semantic/resources/mdtest/binary/integers.md

```diff
@@ -20,28 +20,37 @@ reveal_type(f)  # revealed: Literal[2]
 
 ## Division by Zero
 
-```py
-class MyInt(int):
-    def __truediv__(self, other):
-        return 100
-
-def returns_int() -> int:
-    return MyInt(3)
+This error is really outside the current Python type system, because e.g. `int.__truediv__` and
+friends are not annotated with any special overload for `Literal[0]` to indicate that it's an error,
+and we don't even have a facility to permit such an annotation (the closest would be to annotate the
+overload as returning `typing.NoReturn`, but this just clarifies that it's terminal, it doesn't
+permit any useful diagnostic). So arguably divide-by-zero should be a lint error rather than a type
+checker error. But we choose to go ahead and error in the cases that are very likely to be an error:
+dividing something typed as `int` or `float` by something known to be `Literal[0]`.
+
+This isn't _definitely_ an error, because the object typed as `int` or `float` could be an instance
+of a custom subclass which overrides division behavior to handle zero without error. But if this
+unusual case occurs, the error can be avoided by explicitly typing the dividend as that safe custom
+subclass; we only emit the error if the LHS type is exactly `int` or `float`, not if its a subclass.
 
-# TODO: `a` should be `int` and `e` should be `float` once we support inference.
+```py
 a = 1 / 0  # error: "Cannot divide object of type `Literal[1]` by zero"
 b = 2 // 0  # error: "Cannot floor divide object of type `Literal[2]` by zero"
 c = 3 % 0  # error: "Cannot reduce object of type `Literal[3]` modulo zero"
-# even `int` type could be a subclass of `int` with custom behavior; no error
-d = returns_int() / 0
-# this could be flagged as an error, if we had an ExactFloat or ExactInstance
-# type, but given only a `float` type we can't issue an error for the same
-# reason: could be a custom float subclass
-e = 1.0 / 0
+d = int() / 0  # error: "Cannot divide object of type `int` by zero"
+e = 1.0 / 0  # error: "Cannot divide object of type `float` by zero"
 
 reveal_type(a)  # revealed: float
 reveal_type(b)  # revealed: int
 reveal_type(c)  # revealed: int
+# TODO should be int
 reveal_type(d)  # revealed: @Todo
+# TODO should be float
 reveal_type(e)  # revealed: @Todo
+
+class MyInt(int): pass
+
+# No error for a subclass of int
+# TODO should be float
+reveal_type(MyInt(3) / 0)  # revealed: @Todo
```

Comment:
I think you need to define `MyInt`

Reply:
It's defined four lines above here, on line 51.

It doesn't actually implement `__truediv__` to do anything different. I think that's a distraction, since it's irrelevant to the logic we actually implement, which just always silences the diagnostic for all subclasses.

Reply:
> It's defined four lines above here, on line 51.

🤦 sorry, trying to do too many things at once